### PR TITLE
[Serve] Add Perf Tuning Doc

### PR DESCRIPTION
**Pull request by simon-mo**
This PR adds the performance tuning guide for Ray Serve.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/performance.rst [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+to increase the performance of your deep learning framework?
+
+Batching
+^^^^^^^^
+If your backend can process a batch at a time at a sublinear latency 
+(for example, if it takes 1ms to process 1 query and 5ms to process 10 of them) 
+then batching is your best approach. Check out the :ref:`batching guide <serve-ml-batching>` to 
+make your backend accept batches. You might want to tune your ``max_batch_size`` and ``batch_wait_timeout`` to maximize the benefits:
+
+- ``max_batch_size`` specifies how big the batch should be. Generally,
```
Comment (architkulkarni) [inline doc/source/serve/performance.rst:52]: I can't figure out what's making it bold:  Screen Shot 2021-02-25 at 10 59 01 AM
 Maybe rearrange some things and see if it goes away?

File: doc/source/serve/performance.rst [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+- Ray Serve’s latency overhead is single digit milliseconds, often times just 1-2 milliseconds.
+- For throughput, Serve achieves about 3-4k queries per second on a single machine (8 cores) using 1 http proxy and 8 backend replicas performing noop.
+- It is horizontally scalable so you can add more machines to increase the overall throughput. Ray Serve is built on top of Ray, 
+  so its scalability is bounded by Ray’s scalability. Please check out Ray’s `scalability envelope <https://github.com/ray-project/ray/blob/master/benchmarks/README.md>`_
+  to learn more about the maximum number of nodes and other limitations.
+
+You can check out our `microbenchmark instruction <https://github.com/ray-project/ray/blob/master/python/ray/serve/benchmarks/README.md>`_
+to benchmark on your hardware.
+
+Debug performance Issue
```
Comment (architkulkarni) [inline doc/source/serve/performance.rst:28]:
```suggestion
Debugging performance issues
```

File: doc/source/serve/performance.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+Async functions
+^^^^^^^^^^^^^^^
+Are you using ``async def`` in your callable? If you are using asyncio and
+hitting the same queuing issue mentioned above, you might want to increase 
+``max_concurrent_queries``. Serve sets a low number by default so the client gets 
+proper backpressure. You can increase the value in the :mod:`backend config <ray.serve.config.BackendConfig>`
+to allow more coroutines running in the same replica.
+
+Scale HTTP servers
```
Comment (architkulkarni) [inline doc/source/serve/performance.rst:72]:
```suggestion
Scaling HTTP servers
```

**Review by architkulkarni — CHANGES_REQUESTED** [review]
Looks great!

File: doc/source/serve/performance.rst [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    While this section offers some tips and tricks to improve the performance of your Serve deployment,
+    the :ref:`architecture doc <serve-architecture>` is helpful to gain a deeper understanding of these contexts and parameters.
+
+.. contents::
+
+Performance and known benchmarks
+--------------------------------
+We are continuously benchmarking Ray Serve. The metrics we care about are latency, throughput, and scalability. We can confidently say:
+
+- Ray Serve’s latency overhead is single digit milliseconds, often times just 1-2 milliseconds.
```
Comment (edoakes) [inline doc/source/serve/performance.rst:19]:
```suggestion
- Ray Serve’s latency overhead is single digit milliseconds, around 1-2 milliseconds on average.
```

File: doc/source/serve/performance.rst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+- For throughput, Serve achieves about 3-4k queries per second on a single machine (8 cores) using 1 http proxy and 8 backend replicas performing noop.
```
Comment (edoakes) [inline doc/source/serve/performance.rst:20]:
```suggestion
- For throughput, Serve achieves about 3-4k queries per second on a single machine (8 cores) using 1 http proxy and 8 backend replicas performing noop requests.
```

File: doc/source/serve/performance.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+- It is horizontally scalable so you can add more machines to increase the overall throughput. Ray Serve is built on top of Ray, 
+  so its scalability is bounded by Ray’s scalability. Please check out Ray’s `scalability envelope <https://github.com/ray-project/ray/blob/master/benchmarks/README.md>`_
+  to learn more about the maximum number of nodes and other limitations.
+
+You can check out our `microbenchmark instruction <https://github.com/ray-project/ray/blob/master/python/ray/serve/benchmarks/README.md>`_
+to benchmark on your hardware.
+
+Debug performance Issue
+-----------------------
+A typical performance issue in Serve has the symptom of high client call latency and low throughput.
```
Comment (edoakes) [inline doc/source/serve/performance.rst:30]:
```suggestion
The performance issue you're most likely to encounter is high latency and/or low throughput for requests.
```

File: doc/source/serve/performance.rst [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+This shows the system is processing fewer queries than desired!
```
Comment (edoakes) [inline doc/source/serve/performance.rst:31]: I'd remove this sentence, redundant with the previous

File: doc/source/serve/performance.rst [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+You can check out our `microbenchmark instruction <https://github.com/ray-project/ray/blob/master/python/ray/serve/benchmarks/README.md>`_
+to benchmark on your hardware.
+
+Debug performance Issue
+-----------------------
+A typical performance issue in Serve has the symptom of high client call latency and low throughput.
+This shows the system is processing fewer queries than desired! 
+
+If you have set up :ref:`monitoring <serve-monitoring>` with Ray and Ray Serve, you will observe that
```
Comment (edoakes) [inline doc/source/serve/performance.rst:33]:
```suggestion
If you have set up :ref:`monitoring <serve-monitoring>` with Ray and Ray Serve, you will likely observe that
```

File: doc/source/serve/performance.rst [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Make sure you are using the right hardware and resources. 
+Are you using GPUs (``actor_init_options={“num_gpus”: 1}``) or 1+ cores (``actor_init_options={“num_cpus”: 2}``, and setting ``OMP_NUM_THREADS``)
+to increase the performance of your deep learning framework?
+
+Batching
+^^^^^^^^
+If your backend can process a batch at a time at a sublinear latency 
+(for example, if it takes 1ms to process 1 query and 5ms to process 10 of them) 
+then batching is your best approach. Check out the :ref:`batching guide <serve-ml-batching>` to
```
Comment (edoakes) [inline doc/source/serve/performance.rst:49]: Mention that this is often the case for GPU-based ML inference

File: doc/source/serve/performance.rst (hunk 3 above)
Comment (edoakes) [inline doc/source/serve/performance.rst:72]: Do we have any good way to diagnose that this is the issue? Can we export some metrics to make it clear?

File: doc/source/serve/performance.rst (hunk 1 above)
Comment (simon-mo) [reply doc/source/serve/performance.rst:52]: ah fixed

File: doc/source/serve/performance.rst (hunk 3 above)
Comment (simon-mo) [reply doc/source/serve/performance.rst]: hmm probably just the CPU utilization for the HTTP servers?

**Review by edoakes — APPROVED** [review]

**Review by architkulkarni — APPROVED** [review]

**Pull request merged**